File: src/main/cpp/piecetype.cpp
/*
 * Copyright (C) 2013-2019 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

#include <exception>
#include "piecetype.h"

namespace pulse::piecetype {

bool isValidPromotion(int piecetype) {
	switch (piecetype) {
		case KNIGHT:
		case BISHOP:
		case ROOK:
		case QUEEN:
			return true;
		default:
			return false;
	}
}

bool isSliding(int piecetype) {
	switch (piecetype) {
		case BISHOP:
		case ROOK:
		case QUEEN:
			return true;
		case PAWN:
		case KNIGHT:
		case KING:
			return false;
		default:
			throw std::exception();
	}
}

int getValue(int piecetype) {
	switch (piecetype) {
		case PAWN:
			return PAWN_VALUE;
		case KNIGHT:
			return KNIGHT_VALUE;
		case BISHOP:
			return BISHOP_VALUE;
		case ROOK:
			return ROOK_VALUE;
		case QUEEN:
			return QUEEN_VALUE;
		case KING:
			return KING_VALUE;
		default:
			throw std::exception();
	}
}
}

File: pulse-go/internal/pulse/engine/piece_type.go
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package engine

import "fmt"

type PieceType = int

const (
	Pawn   PieceType = 0
	Knight PieceType = 1
	Bishop PieceType = 2
	Rook   PieceType = 3
	Queen  PieceType = 4
	King   PieceType = 5

	NoPieceType PieceType = 6
)

var (
	PieceTypes = [6]PieceType{Pawn, Knight, Bishop, Rook, Queen, King}
)

func IsValidPieceType(pt PieceType) bool {
	switch pt {
	case Pawn, Knight, Bishop, Rook, Queen, King:
		return true
	default:
		return false
	}
}

func isSliding(pt PieceType) bool {
	switch pt {
	case Bishop, Rook, Queen:
		return true
	case Pawn, Knight, King:
		return false
	default:
		panic(fmt.Sprintf("Invalid piece type: %v", pt))
	}
}

func pieceTypeValueOf(pt PieceType) value {
	switch pt {
	case Pawn:
		return pawnValue
	case Knight:
		return knightValue
	case Bishop:
		return bishopValue
	case Rook:
		return rookValue
	case Queen:
		return queenValue
	case King:
		return kingValue
	default:
		panic(fmt.Sprintf("Invalid piece type: %v", pt))
	}
}
